hpx_topology library should depend on hwloc

## Expected Behavior

Up until a few days ago, building HPX worked fine.

## Actual Behavior

Something has changed in the link order which causes libhwloc.so to come before hpx_topology.a in the link order.  Unfortunately, the hwloc symbols in hpx_topology are unresolved because of it.  If I manually change link.txt to add libhwloc.so at the end of the link line, it works fine.

HPX fails to link libhpx.so (or libhpxd.so for debug builds) : http://ktau.nic.uoregon.edu:8010/#/builders/41/builds/9

## Steps to Reproduce the Problem

... Please be as specific as possible while describing how to reproduce your problem.

  1. cmake -DCMAKE_BUILD_TYPE=Debug -DBOOST_ROOT=/usr/local/boost/1.65.0-gcc6 -DTCMALLOC_ROOT=/usr/local/gperftools/2.5 -DHPX_WITH_MALLOC=tcmalloc -DCMAKE_INSTALL_PREFIX=. -DHPX_WITH_THREAD_IDLE_RATES=OFF -DHPX_WITH_PARCELPORT_MPI=ON -DHPX_WITH_PARCEL_COALESCING=ON -DHPX_WITH_TOOLS=OFF -DHPX_WITH_STACKOVERFLOW_DETECTION_DEFAULT=OFF -DHPX_WITH_APEX=TRUE -DHPX_WITH_APEX_NO_UPDATE=TRUE -DHPX_WITH_APEX_TAG=develop -DAPEX_WITH_BFD=TRUE -DAPEX_WITH_ACTIVEHARMONY=TRUE -DACTIVEHARMONY_ROOT=/usr/local/activeharmony/4.6 -DAPEX_WITH_PLUGINS=TRUE -DAPEX_WITH_PAPI=TRUE -DPAPI_ROOT=/usr/local/papi/5.5.0 -DAPEX_WITH_OTF2=TRUE -DOTF2_ROOT=/usr/local/otf2/2.1 /dev/shm/hpx
  1.  make
  1.  see link fail

## Specifications

... Please describe your environment

  - HPX Version:  current stable
  - Platform (compiler, OS): gcc 6.5 on ubuntu 16, x86_64

The exact error:

```
[ 40%] Linking CXX shared library ../lib/libhpxd.so
../lib/libhpx_topologyd.a(topology.cpp.o): In function `hwloc_get_next_obj_covering_cpuset_by_depth':
/usr/include/hwloc/helper.h:313: undefined reference to `hwloc_bitmap_intersects'
../lib/libhpx_topologyd.a(topology.cpp.o): In function `hwloc_cpuset_to_nodeset_strict':
/usr/include/hwloc/helper.h:952: undefined reference to `hwloc_get_type_depth'
/usr/include/hwloc/helper.h:956: undefined reference to `hwloc_bitmap_zero'
/usr/include/hwloc/helper.h:959: undefined reference to `hwloc_bitmap_set'
../lib/libhpx_topologyd.a(topology.cpp.o): In function `hwloc_cpuset_from_nodeset':
/usr/include/hwloc/helper.h:973: undefined reference to `hwloc_get_type_depth'
/usr/include/hwloc/helper.h:977: undefined reference to `hwloc_bitmap_iszero'
/usr/include/hwloc/helper.h:978: undefined reference to `hwloc_bitmap_zero'
/usr/include/hwloc/helper.h:981: undefined reference to `hwloc_bitmap_fill'
/usr/include/hwloc/helper.h:985: undefined reference to `hwloc_bitmap_zero'
/usr/include/hwloc/helper.h:988: undefined reference to `hwloc_bitmap_isset'
/usr/include/hwloc/helper.h:990: undefined reference to `hwloc_bitmap_or'
../lib/libhpx_topologyd.a(topology.cpp.o): In function `hwloc_get_type_or_below_depth':
/usr/include/hwloc/inlines.h:31: undefined reference to `hwloc_get_type_depth'
/usr/include/hwloc/inlines.h:37: undefined reference to `hwloc_get_type_depth'
/usr/include/hwloc/inlines.h:38: undefined reference to `hwloc_get_depth_type'
/usr/include/hwloc/inlines.h:38: undefined reference to `hwloc_compare_types'
../lib/libhpx_topologyd.a(topology.cpp.o): In function `hwloc_get_nbobjs_by_type':
/usr/include/hwloc/inlines.h:65: undefined reference to `hwloc_get_type_depth'
/usr/include/hwloc/inlines.h:70: undefined reference to `hwloc_get_nbobjs_by_depth'
../lib/libhpx_topologyd.a(topology.cpp.o): In function `hwloc_get_obj_by_type':
/usr/include/hwloc/inlines.h:76: undefined reference to `hwloc_get_type_depth'
/usr/include/hwloc/inlines.h:81: undefined reference to `hwloc_get_obj_by_depth'
../lib/libhpx_topologyd.a(topology.cpp.o): In function `hwloc_get_next_obj_by_depth':
/usr/include/hwloc/inlines.h:88: undefined reference to `hwloc_get_obj_by_depth'
../lib/libhpx_topologyd.a(topology.cpp.o): In function `hpx::threads::operator<<(std::ostream&, hpx::threads::hpx_hwloc_bitmap_wrapper const*)':
/dev/shm/hpx/libs/topology/src/topology.cpp:147: undefined reference to `hwloc_bitmap_snprintf'
../lib/libhpx_topologyd.a(topology.cpp.o): In function `hpx::threads::topology::topology()':
/dev/shm/hpx/libs/topology/src/topology.cpp:207: undefined reference to `hwloc_topology_init'
/dev/shm/hpx/libs/topology/src/topology.cpp:214: undefined reference to `hwloc_topology_load'
../lib/libhpx_topologyd.a(topology.cpp.o): In function `hpx::threads::topology::~topology()':
/dev/shm/hpx/libs/topology/src/topology.cpp:330: undefined reference to `hwloc_topology_destroy'
../lib/libhpx_topologyd.a(topology.cpp.o): In function `hpx::threads::topology::set_thread_affinity_mask(unsigned long, hpx::error_code&) const':
/dev/shm/hpx/libs/topology/src/topology.cpp:453: undefined reference to `hwloc_bitmap_alloc'
/dev/shm/hpx/libs/topology/src/topology.cpp:462: undefined reference to `hwloc_get_obj_by_depth'
/dev/shm/hpx/libs/topology/src/topology.cpp:465: undefined reference to `hwloc_bitmap_set'
/dev/shm/hpx/libs/topology/src/topology.cpp:471: undefined reference to `hwloc_set_cpubind'
/dev/shm/hpx/libs/topology/src/topology.cpp:475: undefined reference to `hwloc_set_cpubind'
/dev/shm/hpx/libs/topology/src/topology.cpp:479: undefined reference to `hwloc_bitmap_snprintf'
/dev/shm/hpx/libs/topology/src/topology.cpp:480: undefined reference to `hwloc_bitmap_free'
/dev/shm/hpx/libs/topology/src/topology.cpp:496: undefined reference to `hwloc_bitmap_free'
../lib/libhpx_topologyd.a(topology.cpp.o): In function `hpx::threads::topology::get_thread_affinity_mask_from_lva(void const*, hpx::error_code&) const':
/dev/shm/hpx/libs/topology/src/topology.cpp:511: undefined reference to `hwloc_bitmap_alloc'
/dev/shm/hpx/libs/topology/src/topology.cpp:521: undefined reference to `hwloc_get_area_membind_nodeset'
/dev/shm/hpx/libs/topology/src/topology.cpp:526: undefined reference to `hwloc_bitmap_alloc'
/dev/shm/hpx/libs/topology/src/topology.cpp:530: undefined reference to `hwloc_bitmap_free'
/dev/shm/hpx/libs/topology/src/topology.cpp:540: undefined reference to `hwloc_get_obj_by_depth'
/dev/shm/hpx/libs/topology/src/topology.cpp:542: undefined reference to `hwloc_bitmap_isset'
/dev/shm/hpx/libs/topology/src/topology.cpp:546: undefined reference to `hwloc_bitmap_free'
../lib/libhpx_topologyd.a(topology.cpp.o): In function `hpx::threads::topology::init_node_number(unsigned long, hwloc_obj_type_t)':
/dev/shm/hpx/libs/topology/src/topology.cpp:618: undefined reference to `hwloc_compare_types'
../lib/libhpx_topologyd.a(topology.cpp.o): In function `hpx::threads::topology::extract_node_mask(hwloc_obj*, unsigned long&) const':
/dev/shm/hpx/libs/topology/src/topology.cpp:640: undefined reference to `hwloc_compare_types'
/dev/shm/hpx/libs/topology/src/topology.cpp:650: undefined reference to `hwloc_compare_types'
../lib/libhpx_topologyd.a(topology.cpp.o): In function `hpx::threads::topology::extract_node_count(hwloc_obj*, hwloc_obj_type_t, unsigned long) const':
/dev/shm/hpx/libs/topology/src/topology.cpp:676: undefined reference to `hwloc_compare_types'
../lib/libhpx_topologyd.a(topology.cpp.o): In function `hpx::threads::topology::cpuset_to_nodeset(unsigned long) const':
/dev/shm/hpx/libs/topology/src/topology.cpp:871: undefined reference to `hwloc_bitmap_alloc'
/dev/shm/hpx/libs/topology/src/topology.cpp:877: undefined reference to `hwloc_bitmap_free'
../lib/libhpx_topologyd.a(topology.cpp.o): In function `hpx::threads::topology::get_cpubind_mask(hpx::error_code&) const':
/dev/shm/hpx/libs/topology/src/topology.cpp:1174: undefined reference to `hwloc_bitmap_alloc'
/dev/shm/hpx/libs/topology/src/topology.cpp:1181: undefined reference to `hwloc_get_cpubind'
/dev/shm/hpx/libs/topology/src/topology.cpp:1183: undefined reference to `hwloc_bitmap_free'
/dev/shm/hpx/libs/topology/src/topology.cpp:1195: undefined reference to `hwloc_get_obj_by_depth'
/dev/shm/hpx/libs/topology/src/topology.cpp:1197: undefined reference to `hwloc_bitmap_isset'
/dev/shm/hpx/libs/topology/src/topology.cpp:1202: undefined reference to `hwloc_bitmap_free'
../lib/libhpx_topologyd.a(topology.cpp.o): In function `hpx::threads::topology::get_cpubind_mask(std::thread&, hpx::error_code&) const':
/dev/shm/hpx/libs/topology/src/topology.cpp:1213: undefined reference to `hwloc_bitmap_alloc'
/dev/shm/hpx/libs/topology/src/topology.cpp:1225: undefined reference to `hwloc_get_thread_cpubind'
/dev/shm/hpx/libs/topology/src/topology.cpp:1229: undefined reference to `hwloc_bitmap_free'
/dev/shm/hpx/libs/topology/src/topology.cpp:1241: undefined reference to `hwloc_get_obj_by_depth'
/dev/shm/hpx/libs/topology/src/topology.cpp:1243: undefined reference to `hwloc_bitmap_isset'
/dev/shm/hpx/libs/topology/src/topology.cpp:1248: undefined reference to `hwloc_bitmap_free'
../lib/libhpx_topologyd.a(topology.cpp.o): In function `hpx::threads::topology::allocate(unsigned long) const':
/dev/shm/hpx/libs/topology/src/topology.cpp:1261: undefined reference to `hwloc_alloc'
../lib/libhpx_topologyd.a(topology.cpp.o): In function `hpx::threads::topology::allocate_membind(unsigned long, std::shared_ptr<hpx::threads::hpx_hwloc_bitmap_wrapper>, hpx::threads::hpx_hwloc_membind_policy, int) const':
/dev/shm/hpx/libs/topology/src/topology.cpp:1277: undefined reference to `hwloc_alloc_membind_nodeset'
../lib/libhpx_topologyd.a(topology.cpp.o): In function `hpx::threads::topology::set_area_membind_nodeset(void const*, unsigned long, void*) const':
/dev/shm/hpx/libs/topology/src/topology.cpp:1292: undefined reference to `hwloc_set_area_membind_nodeset'
../lib/libhpx_topologyd.a(topology.cpp.o): In function `hpx::threads::topology::get_area_membind_nodeset(void const*, unsigned long) const':
/dev/shm/hpx/libs/topology/src/topology.cpp:1326: undefined reference to `hwloc_bitmap_alloc'
/dev/shm/hpx/libs/topology/src/topology.cpp:1338: undefined reference to `hwloc_get_area_membind_nodeset'
../lib/libhpx_topologyd.a(topology.cpp.o): In function `hpx::threads::topology::deallocate(void*, unsigned long) const':
/dev/shm/hpx/libs/topology/src/topology.cpp:1382: undefined reference to `hwloc_free'
../lib/libhpx_topologyd.a(topology.cpp.o): In function `hpx::threads::topology::mask_to_bitmap(unsigned long, hwloc_obj_type_t) const':
/dev/shm/hpx/libs/topology/src/topology.cpp:1389: undefined reference to `hwloc_bitmap_alloc'
/dev/shm/hpx/libs/topology/src/topology.cpp:1390: undefined reference to `hwloc_bitmap_zero'
/dev/shm/hpx/libs/topology/src/topology.cpp:1399: undefined reference to `hwloc_get_obj_by_depth'
/dev/shm/hpx/libs/topology/src/topology.cpp:1402: undefined reference to `hwloc_bitmap_set'
../lib/libhpx_topologyd.a(topology.cpp.o): In function `hpx::threads::topology::bitmap_to_mask(hwloc_bitmap_s*, hwloc_obj_type_t) const':
/dev/shm/hpx/libs/topology/src/topology.cpp:1420: undefined reference to `hwloc_get_obj_by_depth'
/dev/shm/hpx/libs/topology/src/topology.cpp:1422: undefined reference to `hwloc_bitmap_isset'
../lib/libhpx_topologyd.a(topology.cpp.o): In function `hpx::threads::hpx_hwloc_bitmap_wrapper::~hpx_hwloc_bitmap_wrapper()':
/dev/shm/hpx/libs/topology/include/hpx/topology/topology.hpp:51: undefined reference to `hwloc_bitmap_free'
../lib/libhpx_topologyd.a(topology.cpp.o): In function `hpx::threads::hpx_hwloc_bitmap_wrapper::reset(hwloc_bitmap_s*)':
/dev/shm/hpx/libs/topology/include/hpx/topology/topology.hpp:57: undefined reference to `hwloc_bitmap_free'
collect2: error: ld returned 1 exit status
src/CMakeFiles/hpx.dir/build.make:4620: recipe for target 'lib/libhpxd.so.1.4.0' failed
make[2]: *** [lib/libhpxd.so.1.4.0] Error 1
CMakeFiles/Makefile2:35478: recipe for target 'src/CMakeFiles/hpx.dir/all' failed
make[1]: *** [src/CMakeFiles/hpx.dir/all] Error 2
Makefile:140: recipe for target 'all' failed
make: *** [all] Error 2
program finished with exit code 2
```

--- a/libs/topology/CMakeLists.txt
+++ b/libs/topology/CMakeLists.txt
@@ -25,6 +25,11 @@ set(topology_sources
   topology.cpp
 )
 
+find_package(Hwloc QUIET)
+if(NOT HWLOC_FOUND)
+  hpx_error("    Hwloc could not be found, please specify HWLOC_ROOT to point to the correct location")
+endif()
+
 include(HPX_AddModule)
 add_hpx_module(topology
   COMPATIBILITY_HEADERS ON    # Added in 1.4.0
@@ -42,6 +47,7 @@ add_hpx_module(topology
     hpx_format
     hpx_logging
     hpx_type_support
+    ${HWLOC_LIBRARIES}
   CMAKE_SUBDIRS examples tests
 )
 

